Move Guide topic: Garbage Collection.

steveperry-53 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2488)
<!-- Reviewable:end -->

chenopis commented on docs/concepts/abstractions/controllers/garbage-collection.md:
```diff
@@ -0,0 +1,70 @@
+
+If you specify `deleteOptions.orphanDependents=true`, or leave it blank,
+then the garbage collector first resets the `ownerReferences` in the
+dependents, and then delete the owner. In this case, the deletion of the owner
```
typo: "delete" -> "deletes"

caesarxuchao replied:
and then deletes **only** the owner, but not its dependents.

chenopis commented on docs/concepts/abstractions/controllers/garbage-collection.md:
```diff
@@ -0,0 +1,70 @@
+
+{% capture whatsnext %}
+
+[Design Doc](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/garbage-collection.md)
```
Maybe point this link directly to https://github.com/kubernetes/community/blob/master/contributors/design-proposals/garbage-collection.md

steveperry-53 commented:
@pwittrock, Can you suggest a technical reviewer for this. I'd especially like someone to take a look at the section on Controlling whether the garbage collector deletes dependents.

pwittrock commented:
@caesarxuchao Would you take a look at this?

pwittrock commented:
@steveperry-53 Garbage collection is a meta topic about Controllers, vs a Concept guide about a particular controller.  Do you think we should logically separate the 2 in the ToC?

pwittrock commented on docs/concepts/abstractions/controllers/garbage-collection.md:
````diff
@@ -0,0 +1,70 @@
+```
+
+If you specify `deleteOptions.orphanDependents=true`, or leave it blank,
+then the garbage collector first resets the `ownerReferences` in the
````
what does "resets" mean here?  Why is this important?

caesarxuchao replied:
It means "removes". I'm trying to explain why the deletion of the owner is asynchronous. Maybe the text is clearer this way:

If you specify `deleteOptions.orphanDependents=true`, or leave it blank, the object will be deleted asynchronously, while its dependents will not be garbage collected. "Asynchronously" means that the API server returns s a 200 OK in respond to the delete request before the object gets deleted, because the garbage collector needs to remove the `ownerReferences` in the dependents in the background before the object can be safely deleted.

pwittrock commented on docs/concepts/abstractions/controllers/garbage-collection.md:
```diff
@@ -0,0 +1,70 @@
+specify relationships between owners and dependents by manually setting the
+`ownerReference` field.
+
+## Controlling whether the garbage collector deletes dependents
```
This title seems misleading.  The content implies that the garbage collector always deletes dependents, either synchronously or asynchronously.

Maybe

> ## Configuring garbage collection order
>
> By default, garbage collection will run against the dependent objects before the owning object is deleted.  When run this way, the deletion of both the owner and is dependents are asynchronous, resulting in a 200 OK response before completion.
> Alternatively, the owning object can be synchronously before running garbage collection by specifying `deleteOptions.orphanDependents=false`.  This will guarantee that the owning object is deleted before the 200 OK response is returned, however the dependent objects are deleted asynchronously after the owner.

caesarxuchao replied:
Sorry, its not the title is misleading, it's the description. OrphanDependents means "no garbage collection".

pwittrock commented on docs/concepts/abstractions/controllers/garbage-collection.md:
```diff
@@ -0,0 +1,70 @@
+Some Kubernetes objects are owners of other objects. For example, a ReplicaSet
+is the owner of a set of Pods. The owned objects are called *dependents* of the
+owner object. Every dependent object has a `metadata.ownerReferences` field that
+points to the owning API object.
```
Would be nice to inline an example Pod / Deployment with this set.

caesarxuchao replied:
An excerpt of a pod with an ownerReference pointing to a replicaset:
```
$ kubectl get pod example-123456789-12345 -o yaml
apiVersion: v1
kind: Pod
metadata:
  ownerReferences:
  - apiVersion: extensions/v1beta1
    controller: true
    kind: ReplicaSet
    name: example-123456789
    uid: 12345678-1234-1234-1234-123456789012
...
```

steveperry-53 commented:
@caesarxuchao, @pwittrock, @chenopis, I've addressed most of your comments.

@pwittrock, To answer your question about where this should go in the TOC: Let's leave it here for now. Soon we need to do a full revision of the Concepts TOC, but let's save that for a separate PR. @chenopis, Let's discuss.

chenopis commented:
@steveperry-53 Ok, yeah, let's discuss the Concepts ToC. Feel free to put something on my calendar. Cheers

pwittrock commented:
lgtm
